### PR TITLE
Add support for PayPal JS SDK

### LINKED ISSUE #626: Master issue: Update deprecated `checkout.js` script
PayPal has announced that they will be deprecating the [`checkout.js` script](https://developer.paypal.com/docs/archive/checkout/) we use in [our code](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/9487b8d757cc2ea5f063bb7c534e198beaa5b651/includes/class-wc-gateway-ppec-cart-handler.php#L461). Their documentation includes [instructions on how to update our code base](https://developer.paypal.com/docs/checkout/reference/upgrade-integration/#6-fix-deprecations) to use their new SDK.

>If you are using a previous version of the PayPal Checkout integration, including custom buttons, full-page redirects, or `checkout.js`, PayPal recommends that you upgrade to the latest version of the integration.
>
> [Source](https://developer.paypal.com/docs/checkout/reference/).

As far as I can tell, PayPal have not specified a date for when the `checkout.js` script will no longer be available so I'm not sure how much time we have to update the script; I assume it's in the order of months.

>Checkout with `checkout.js`. Valid before February 2019. Customers who use `checkout.js` can access reference and support material in this Checkout guide. However, PayPal does not update `checkout.js` with new features and enhancements.
>
> [Source](https://developer.paypal.com/docs/archive/checkout/).

At the very least `checkout.js` will not be receiving any updates; it was archived in February 2019, almost 8 months ago.

This seems to at the very least change how smart buttons function [[#](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/622#issuecomment-537692208)], so we will need to update documentation as we deploy the update. Other features have not been tested for changes as of the writing of this issue.

To me, these are the most notable changes:

- Passing `client-id` to the CDN when loading the SDK, instead of passing `env` and `client` during the component `render()` call.
- Passing `commit` to the CDN when loading the SDK, instead of passing `commit` during the component `render()` call.
- Passing `locale` to the CDN when loading the SDK, instead of passing `locale` during the component `render()` call.
- Use vanilla JS `promise` and `fetch` instead of `paypal.Promise` and `paypal.request`. This has some implications regarding IE11 compatibility [[#](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/622#discussion_r330920096)]

<hr />

# Child issues

- [ ] Store `client-id` so we can load the new SDK (#645).
- [ ] Pass `disable-funding` and `disable-card` when we load the new SDK (#647).
- [ ] Pass `commit` when we load the new SDK (#648).
- [ ] Add polyfills for `fetch` and `Promise` (#649).
- [ ] Redirect to a cancellation page when appropriate (#650).
- [ ] Catch and handle errors that happen during the checkout flow (#651).

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Description

**Work in progress.** Intended to supersede #622.

This PR adds support for the new [JS SDK integration](https://developer.paypal.com/docs/checkout/) for Smart Payment Buttons.
It doesn't directly replace the current implementation (which uses [`checkout.js`](https://developer.paypal.com/docs/archive/checkout/)). Instead, depending on whether the REST API credentials are available or not, it uses either the JS SDK or the old implementation.

The code also tries to appropriately map properties/functionality from the [old implementation to the new one](https://developer.paypal.com/docs/checkout/reference/upgrade-integration).

### To get your Client ID and secret
1. Go to https://developer.paypal.com/developer/applications/ and log into your sandbox developer account
2. Go to **My Apps & Credentials** from the nav menu.
3. Under **REST API apps**, click **Create App**.
4. Give it a name and drop the dropdown select the merchant sandbox credentials you use on your site. 
5. Copy the Client ID
6. Click _Show_ under **Secret** and copy it. 

Docs: https://developer.paypal.com/docs/checkout/integrate/#1-set-up-your-development-environment

### Steps to test
1. Get the REST API credentials (Client ID and secret) from your PayPal dashboard.
2. Go to the plugin settings:
2.1. Enter the live/sandbox REST API credentials.
2.2. Make sure the "Smart Payment Buttons" setting is enabled.
2.3. For testing purposes, enable PayPal Checkout for the single product page and cart page.
3. Make sure the PayPal button appears as configured, is clickable and completes the payment process for each of these scenarios:

    <details>
    <summary>Test Case: Regular checkout process</summary>
      1. Add an item to the cart.<br />
      2. Go to the cart page.<br />
      3. Go to the checkout page.<br />
      4. Use the PayPal button to complete the payment process.
   </details>

    <details>
    <summary>Test Case: Pay directly from a product's page</summary>
      1. Open an item's product page.<br />
      2. Use the PayPal button to complete the payment process.
   </details>

    <details>
    <summary>Test Case: Pay directly from the cart page</summary>
      1. Add an item to the cart.<br />
      2. Go to the cart page.
      3. Use the PayPal button to complete the payment process.
   </details>

Closes #626.

### CODE REVIEW

**Comment by jorgeatorres** [comment]
Adding this comment to track to-do's (`pb0GrA-ow-p2`):

#### To Do
- [ ] Add validation for REST API credentials.
- [ ] Add an admin notice (on WC > Settings > Payments) asking customers to enter REST API creds to get the new buttons.
- [ ] Add polyfills for `fetch()` and `Promise` [IE compat] #649
- [ ] Use `onCancel()` to properly handle payment cancellation  #650
- [ ] Handle errors with `onError()` #651
- [ ] Adjust settings when in REST API mode:
  - [ ] Hide checkout, product page, mini-cart specific funding method settings.
  - [ ] Hide "size" setting.
  - [ ] As PayPal Credit is now automatically enabled (depending on country availability, etc.) change the setting to "Enable PayPal Credit to elegible customers"

**Comment by mattallan** [comment]
Hey @jorgeatorres, I'm going to try focus on getting this PR over the line this week so that we can have 1-2 weeks of testing before 1.7 is due to release.

I'll start with testing the current changes but I wanted to check if there's anything left on the Todos that I can help with coding? :)

**Comment by jorgeatorres** [comment]
👋 @mattallan:

> Hey @jorgeatorres, I'm going to try focus on getting this PR over the line this week so that we can have 1-2 weeks of testing before 1.7 is due to release.

Thanks. This is also going to be my focus for the week.

> I'll start with testing the current changes […]

The current changes are missing some commits that I haven't pushed yet. For instance, I added all the funding methods to the backend but I'm still not properly using those on the frontend. The branch also needs a rebase to catch up with `1.7_dev`.
Most of those things I have addressed on my local branch, so I just need to do some more testing and push the changes. I can do that tomorrow (as soon as I'm back from my time off). If it's ok with you, I'll let you know when it's done and we can take it from there.

> […] but I wanted to check if there's anything left on the Todos that I can help with coding? :)

Yes. The remaining todo from above is something I have only looked at superficially:

_Add polyfills for fetch() and Promise [IE compat] #649_

I haven't written any code for that as I'm unsure whether we can just include any polyfill from a 3rd party (as long as it's free software) or not. And if there are several libraries providing the functionality, which one do we choose? If you could do some research on this and/or add the relevant JS code to the plugin, that would be great, that is, if it's something you find interesting. Let me know!

**Comment by mattallan** [comment]
> If it's ok with you, I'll let you know when it's done and we can take it from there.

No worries :) Sounds good to me!

> If you could do some research on this and/or add the relevant JS code to the plugin, that would be great

Yep! I'm happy to start looking into this today and see what I can find :)

**Comment by james-allan** [comment]
I just rebased the last commit to fix an error I received locally. 

```
PHP Fatal error:  Uncaught ParseError: syntax error, unexpected ')' in /app/public/wp-content/plugins/woocommerce-gateway-paypal-express-checkout/includes/class-wc-gateway-ppec-cart-handler.php:528
```

There was a comma at the end of this [multiline function call](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/commit/5b02a5362212c78d39970ca064a5a2057de6e225#diff-c5271c5e9f4b2ddfc403c66a5d3c5375R527).

**Comment by mattallan** [comment]
Is anyone else noticing there's no way to go back to choose a different payment option with this new SDK after clicking Debit credit cards?

On master, there's a _Payment Options_ link: https://d.pr/i/1P0d7J
On this PR, there's no way of getting back: https://d.pr/i/FfFesM

Am I missing something? I can see some customers being unhappy with this change so I wonder if there's something we can do about it :thinking:

**Comment by james-allan** [comment]
Yeah, I can see that too. That whole section is handled by PayPal right so we have limited control over that. One option would be to add our own link or button that refreshes the form or something.

**Comment by mattallan** [comment]
@jorgeatorres I'm getting the following console errors on the checkout page when I try to click on the PPEC credit card button when no billing fields have been entered yet: https://d.pr/i/gHH6Uc

Steps to reproduce:
1. open store in an incognito window (or be signed out/guest customer)
2. add a product to the cart
3. visit checkout
4. select PPEC payment option
5. click the smart button to Pay with Debit Card (should see errors)

Other things I noticed:
 - if you click the smart button again, a different checkout notice appears with "Billing First Name is a required field"
 - if you fill in all the fields and try again it still errors.
 - refreshing the checkout page after filling in all the fields fixes the issue

> That whole section is handled by PayPal right so we have limited control over that.

Ahh yeah true. I guess we can example what our options are in a future release since it's out of our control and priority is getting this tested/merged.

---

@jorgeatorres I'm going to continue to work on testing and the implementing fetch/promise polyfills, but is there something James can help with coding?

I was thinking maybe the changes to the settings page? (i.e. removing the cart/checkout/product SPB options when using the new SDK).

**Comment by jorgeatorres** [comment]
@mattallan: I'm going to look into the errors you've mentioned.

> I was thinking maybe the changes to the settings page? (i.e. removing the cart/checkout/product SPB options when using the new SDK).

This should be already working. If you enter REST credentials, the settings page should automatically change to hide the cart/checkout/product SPB settings that we can no longer support (which I believe was "hide funding methods" and the size setting). We still support button layout/label settings as those can still be configured per context even with the new SDK, and that's why context-specific still exist.

If this is not what you're seeing, please let me know.

**Comment by mattallan** [comment]
> We still support the button layout/label settings as those can still be configured per context even with the new SDK, and that's why context-specific sections still exist.

Ohh nice sorry I didn't realise this was the case :+1: That looks good then

**Comment by james-allan** [comment]
I've been doing some testing of the new SPBs and all the buttons work for simple products using the CC form and PayPal modal on all screens. However, when I started testing subscription products, the CC form doesn't load. 

In the console there aren't any errors, the loading spinner just sits there forever. Can anyone else replicate that? Using the PayPal popup seems to be working just fine.

**Comment by mattallan** [comment]
I'm seeing the same thing. But I'm questioning what is expected behaviour here 🤔 

I think when you purchase a subscription product with PayPal Checkout you aren't allowed to checkout as a guest and you have to create an account with PayPal to purchase subscriptions. 

When I try to purchase a subscription on master with the visa CC fields it comes up with an error, clicking try again opens the PayPal payment window and asks me to create an account: https://d.pr/i/QzKPtD

I wonder if we can just drop the credit card fields if the cart contains a subscription?

**Comment by james-allan** [comment]
> I wonder if we can just drop the credit card fields if the cart contains a subscription?

Yeah that sounds like a good idea given the guest checkout restrictions.

**Comment by james-allan** [comment]
> When I try to purchase a subscription on master with the visa CC fields it comes up with an error, clicking try again opens the PayPal payment window and asks me to create an account: 

Since this issue isn't exclusively related to the new SDK, I think I'll work on that separately, keeping in mind that this also needs to work with this branch and the changes here.

**Comment by mattallan** [comment]
So I have pushed up the polyfill changes on branch [`new_sdk_with_polyfill`](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/compare/update_javascript_sdk...new_sdk_with_polyfills) which adds support for IE 11.

I'm not really sold on the approach but I haven't added polyfill before so I don't know if this is normal 🤔

I noticed [WP 5.0+ includes all the polyfill scripts](https://developer.wordpress.org/block-editor/contributors/develop/scripts/#polyfill-scripts) that we could just enqueue, however, we need to support WP 4.4+ so the only solution I could think of is just including them in our /assets/JS

Thoughts?

If we're okay with this. I'll tidy up the commits and move them into this PR

**Comment by jorgeatorres** [comment]
@mattallan:

>I'm not really sold on the approach but I haven't added polyfill before so I don't know if this is normal 🤔

Looks good to me. AFAIK polyfills are usually included on the page after checking whether the functionality you need is available or not (through JS code) and not via browser detection. That's how [WordPress does it](https://github.com/WordPress/WordPress/blob/58d5aad2a3cccce7f0695a4686c719a22e5b5adc/wp-includes/script-loader.php#L121-L135) with the [`wp_get_script_polyfill()`](https://github.com/WordPress/WordPress/blob/58d5aad2a3cccce7f0695a4686c719a22e5b5adc/wp-includes/script-loader.php#L179) function. That said, `wp_get_script_polyfill()` as well as the polyfills themselves are only available on WP 5.0+ so we can't use that.

In this particular case, though, I think it's fine to rely on `$is_IE` because, among the browsers we support, we literally only need to do this for IE.

> I noticed WP 5.0+ includes all the polyfill scripts that we could just enqueue, however, we need to support WP 4.4+ so the only solution I could think of is just including them in our /assets/JS
> Thoughts?

I think that's fine. My only suggestion would be that we use a minified version (.min.js) of those assets (if available).

Please let me know what you think.

**Comment by mattallan** [comment]
Thanks @jorgeatorres for your notes and feedback. I've pushed up the polyfill changes and have been testing your latest fixes which address https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/668#issuecomment-604156085

The error messages are definitely better but I feel the behaviour/UX is still off and feels broken. I'm not sure if there's a solution to this as the PayPal developer docs aren't clear for what we should be doing.

Basically, if you get an error on the checkout, the PayPal card form never restarts/resets and you have to refresh the page to re-render the buttons.

Steps to reproduce:
1. add a product to your cart
2. go to the checkout and don't fill in a required field
3. click the Debit and Credit card Paypal option
4. Note the checkout error saying "Billing First Name is a required field"
5. Fix up the problem by filling in the required field
6. Notice the PayPal option is not there (because we clicked Debit and Credit Card before)
6. Try click the Debit and Credit card PayPal option again
7. No errors on checkout however CC fields never load and the console errors persist

I've been trying to find a way to restart the form but can't manage to find a solution to this.
Am I missing something?

Is this all related to the issue you mentioned in slack:

> no clean way to cancel promises from within the createOrder callback and the console errors will always be generated

If so 😿

**Comment by mattallan** [comment]
I've very much hit a roadblock with the issue I mentioned above in: https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/668#issuecomment-604844596 😢 

Here's a recording of the issue I'm referring to: https://d.pr/v/Kj4wgP

tl;dr if you leave a required field empty on the checkout and try to purchase with the credit card smart buttons option, you get an unhandled error and you can't checkout, even after fixing up the problem.

The customer has to refresh the page to fix it.

While digging for answers, I found these [docs on implementing validation](https://developer.paypal.com/docs/checkout/integration-features/validation/#asynchronous-validation) but this approach also doesn't fully fix the problem but there were some improvements.

I implemented a basic version of this for testing:

```
onClick: function(data, actions) {
	if ( $( '#billing_first_name').val() == '' ) {
		return actions.reject();
	} else {
		return actions.resolve();
	}
},
```

What I found is after you call `action.reject()`, the Paypal option doesn't go away but the card fields on the checkout never load, even after calling resolve... I was logging various steps in the payment/createOrder function and I noticed that a payment token was generated, but the fields never came back.

One good thing about this approach was the PayPal Button is still there and works perfectly.

Here's a recording of the checkout using the onClick method: https://d.pr/v/pYBWdz

Keen to get your thoughts on this @jorgeatorres and whether there's a better solution to this problem 😄 Thanks!

**Comment by jorgeatorres** [comment]
> tl;dr if you leave a required field empty on the checkout and try to purchase with the credit card smart buttons option, you get an unhandled error and you can't checkout, even after fixing up the problem.

I've been able to reproduce this but not consistently: sometimes the PayPal button sticks around, sometimes it disappears. I haven't been able to pinpoint the reason why.
Also, keep in mind that this happens even if no error occurs during checkout. If you click "Debit or Credit Card" you can't go back and choose PayPal or any other button. This is obviously more critical when an error occurs, but it's not tied to that particular situation. I did some research and [found an open issue about this ](https://github.com/paypal/paypal-checkout-components/issues/1283) in the PayPal Checkout Components repo, which makes me think we shouldn't be trying to work around this if it's ok to wait for PayPal to fix the problem. What are your thoughts on this?

> While digging for answers, I found these docs on implementing validation but this approach also doesn't fully fix the problem but there were some improvements.

Yes, I came across `onClick()` last week when working on the validation code. I don't think `onClick()` was meant for this kind of validation, though. `onClick()` is the perfect place to perform validation __before__ attempting to create an order and obtain its token, but our AJAX call combines both validation and generating the token. It would be difficult to separate these two aspects, but even if we could do it, a different error could still be produced as part of WC's checkout process (not necessarily an empty field error) and we would face the exact same issue: the customer unable to choose a different payment button.

That's why I think this should be addressed on PayPal's side. Also, `onClick()` doesn't work perfectly, as resolving the promise doesn't render the fields again consistently 🤷‍♂, which is something you've experienced yourself.

> Keen to get your thoughts on this @jorgeatorres and whether there's a better solution to this problem […]

I'm not sure if it's actually "a better" solution but if we definitely want to include a fix for this instead of waiting for PayPal to address it, one option would be to re-render the buttons when an error occurs.
For instance, changing the `onError()` handler to something like this:

```js
onError: function() {
	jQuery(selector).empty();
	render();
}
```

Due to the way the current JS code is implemented, this would actually re-render buttons in all locations (mini-cart, etc.) but with some small refactoring we can make it perform better. Is this something you think we should do? If so, I can definitely help with that.

**Comment by james-allan** [comment]
> If you click "Debit or Credit Card" you can't go back and choose PayPal or any other button... I did some research and found an open issue about this in the PayPal Checkout Components repo, which makes me think we shouldn't be trying to work around this if it's ok to wait for PayPal to fix the problem. What are your thoughts on this?

Good find. I agree. Given PayPal are working on it, I don't think we should try to build a workaround. 

> a different error could still be produced as part of WC's checkout process (not necessarily an empty field error) and we would face the exact same issue: the customer unable to choose a different payment button. That's why I think this should be addressed on PayPal's side.

I'll let @mattallan chime in on this but from what I understood the issue isn't that they can't choose a different payment gateway, it's that after fixing a validation error (like a missing field or another validation issue) the customer can't proceed with debit or credit fields. That's not something PayPal can fix, is it? Given it's specific to WC's checkout validation and error handling.

**Comment by jorgeatorres** [comment]
> I'll let @mattallan chime in on this but from what I understood the issue isn't that they can't choose a different payment gateway, it's that after fixing a validation error (like a missing field or another validation issue) the customer can't proceed with debit or credit fields.

Yes, I didn't mean a different gateway but just a different PayPal button, but you're right… the credit/debit card fields aren't reset and clicking the button "does nothing" after an error.

> That's not something PayPal can fix, is it? Given it's specific to WC's checkout validation and error handling.

Even if we ignore WC for a moment, I think it's perfectly "normal" for a remote request during `createOrder()` to fail (for instance, due to server error) and it doesn't make sense -at least to me- that you're left stuck with a PayPal button that no longer works: you can't go back to choose a different funding source and you can't retry with debit/credit. That's why I think this is something they should fix.

It's also not the case that we're doing weird things or abusing the library. Setting up a transaction via a remote request is [a scenario the SDK supports](https://developer.paypal.com/docs/checkout/reference/server-integration/set-up-transaction/).

Having said that, I'm not sure if we can rely on PayPal fixing this (soon or at all). Re-rendering the buttons, while not ideal, would be a possible fix or a way to "reset" the buttons.

**Comment by mattallan** [comment]
> from what I understood the issue isn't that they can't choose a different payment gateway, it's that after fixing a validation error (like a missing field or another validation issue) the customer can't proceed

Yeah so there are a few issues being mentioned here so I'll try to clarify:

1. Not being able to go back and choose Paypal after choosing Debit/credit Cards

This will be fixed by PayPal as part of https://github.com/paypal/paypal-checkout-components/issues/1283 and not something we should fix ourselves

2. Initial Checkout validation errors (like submitting the checkout with a missing required field) stop customers from proceeding

Seen in https://d.pr/v/Kj4wgP

This is the issue I was referring to and I just tested the code @jorge mentioned above and that fixes the issues from what I can see (still need to do more testing) and IMO is the best UX we will be able to get:

```
onError: function() {
	jQuery(selector).empty();
	render();
}
```

3. Some other checkout/order processing error or PayPal API error

As Jorge mentioned, trying to fix this in the `onClick` function will mean it'll still break if there's an issue on PayPal side.

Re-rendering the buttons on `onError` will fix this issue as well.

**Comment by mattallan** [comment]
@jorgeatorres I believe this is all looking really good !!

I just pushed up two commits:
- a658388 fixes the typo found on the settings page
- 4fcf5e7 re-renders the SPB when an error is caught (I think this is the best solution we have to the problem seen in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/668#issuecomment-604844596

One issue I found is while logged in, the PayPal Credit button is not loading on Product and Cart pages, but it does load on the Checkout 🤔 https://d.pr/i/3iJ2fR 

I made sure that PayPal Credit was enabled on those pages but yeah I can only get it to load on the checkout.

If I open in Incognito the product page and cart page both have the PayPal Credit option: https://d.pr/i/sUVBpU

Has this got something to do with the script requiring the customers Country? Or something else? going on?

---

**Testing new SDK changes**

I still need to do some testing using the legacy filter

- **Checkout Page**
	- [x] Check PayPal Checkout is only displayed on checkout when "Enable the PayPal Mark on regular checkout" setting is enabled
	- [x] Checkout specific context settings are used on the checkout
	- [x] Check global SPB settings used when no specific checkout page settings
	- [x] Purchasing product with PayPal button option
	- [x] Purchasing product with Debit Card option
		- Invalid test card (only accepts sandbox test cards)
	- [x]  Use an invalid credit card to fill in the PayPal Debit Card fields
	- [x] checked PP Credit is only visible if not hidden in funding methods and/or "Enable PayPal Credit to eligible customers" is enabled.
	- [x] Purchasing a product with PayPal Credit
		- I was able to test by changing VPN to US: https://d.pr/i/iIBlJw
	- [x] Test Hiding specific finding methods
	- [x] Purchase with Coupon
	- [x] Invalid checkout form fields with when trying to purchase with PayPal (missing required field)
	- [x] Invalid checkout form fields when trying to purchase with the Debit Card option
		- this is broken without 4fcf5e7
	- [x] Invalid checkout fields when trying to purchase with PayPal Credit option
	- [x] Cancel a payment at PayPal and make sure we're redirected back to the cart page
	- [ ] Test checkout purchase with IE 11

- **Cart Page**
	- [x] SPB not displayed when "Enable PayPal Checkout on the cart page" is disabled
	- [x] Purchase with PayPal
	- [x] Purchase with Debit Card
	- [ ] Purchase with PayPal Credit :x: not loading while logged in
	- [x] Customer is redirected to "Confirm your PayPal Order" page

- **Product Page**
	- [x] SPB not displayed when "Checkout on Single Product" is disabled
	- [x] Product page specific settings used only on product page and not in mini-cart
	- [x] Purchase with PayPal
	- [x] Purchase with Debit Card
	- [ ] Purchase with PayPal Credit :x: not loading while logged in
	- [x] Customer is redirected to "Confirm your PayPal Order" page

- **Mini-Cart**
	- [x] SPB not displayed when "Enable PayPal Checkout on the cart page" is disabled
	- [x] Mini-cart specific settings used on mini-cart
	- [x] Purchase with PayPal
	- [x] Purchase with Debit Card
	- [ ] Purchase with PayPal Credit :x: not loading while logged in
	- [x] Customer is redirected to "Confirm your PayPal Order" page

- **Merchant**
	- [x] Check orders in the merchant dashboard

**Comment by jorgeatorres** [comment]
@mattallan: Thanks doing all this testing!

> Has this got something to do with the script requiring the customers Country? Or something else? going on?

To be honest, I'm not sure what the reason for that might be. In my case, I see the Credit option on all pages (when pointing my VPN to a US server). The PayPal script uses the locale, but not the country per se. I don't know how being logged in vs incognito affects how PayPal decides what to display (it might also be related to space somehow?), but I don't know if we have control over that:

From https://developer.paypal.com/docs/checkout/integration-features/customize-button/#which-buttons-will-i-see:
> The buttons that display are decided automatically, based on a range of factors, including:
>
> Buyer country
> Device type
> Funding sources the buyer has opted to see.
> As a result, each buyer might see a unique combination of buttons.

And from https://developer.paypal.com/docs/checkout/integration-features/standalone-buttons/:
> Even with standalone buttons, your integrations take advantage of the smart eligibility logic the PayPal JavaScript SDK provides, meaning only the appropriate buttons for the current buyer automatically display.

I'm not sure if we can force the SDK to display this particular method. In legacy mode, there's a property that allows us to "tell" PayPal explicitly that we want to display PayPal Credit, which probably overrides other considerations, but in SDK mode we no longer have that possibility. We only have control over which methods we don't want to display.

One experiment we could try is this:  could you use the Console in your DevTools and run the following JS command in both scenarios? What is the output?

```js
paypal.Buttons({fundingSource: paypal.FUNDING.CREDIT}).isEligible();
```

**Comment by mattallan** [comment]
Thanks @jorgeatorres for the notes! So when I turned the computer on this morning and loaded the same product page while logged in, the PayPal Credit button is loading properly...

I have no idea what was causing it to not load yesterday but I'll keep an eye out for it incase it reappears :( sorry for the false alarm

>  In legacy mode, there's a property that allows us to "tell" PayPal explicitly that we want to display PayPal Credit, which probably overrides other considerations, but in SDK mode we no longer have that possibility. We only have control over which methods we don't want to display.

Yeah, I think this is actually a good feature of the new SDK.

> One experiment we could try is this

That's handy to know! thanks :)

---

Todos left on this PR:
 - [ ] Remove client/secret settings fields
 - [ ] Set a Woo Client ID
 - [ ] Moar testing

**Comment by jorgeatorres** [comment]
@mattallan: This branch should be in good shape again.

**Comment by mattallan** [comment]
It's happening!!!!

Thanks so much @jorgeatorres for working on this monster!
I spent a bit of time testing this today and I believe I've tested all the critical gateway functions with both the new SDK and also the legacy PayPal Checkout JS (which can be enabled through the filter)

I'm going to merge this one so we can address last remaining todos' in a smaller PR.

- [ ] Remove client/secret settings fields
- [ ] Set a Woo Client ID

Massive kudos and awesome work on this one 💯 🥇 
---

### Testing on the latest rebase

#### Checkout Page

- [x] Check PayPal Checkout is only displayed on checkout when "Enable the PayPal Mark on regular checkout" setting is enabled
  - [x] working with legacy mode enabled
- [x] Checkout specific context settings are used on the checkout
  - [x] working with legacy mode enabled
- [x] Check global SPB settings used when no specific checkout page settings
  - [x] working with legacy mode enabled
- [x] Purchasing product with PayPal button option
  - [x] working with legacy mode enabled
- [x] Purchasing product with Debit Card option
  - [x] working with legacy mode enabled
- [x] Use an invalid credit card to fill in the PayPal Debit Card fields
- [x] Check PP Credit is only visible if not hidden in funding methods and/or "Enable PayPal Credit to eligible customers" is enabled (need to test with VPN set to the US)
  - [x]  working with legacy mode enabled
- [x] Purchasing a product with PayPal Credit
  - [x] working with legacy mode enabled
- [x] Test Hiding specific finding methods
  - [x]  working with legacy mode enabled
- [x] Purchase with Coupon
- [x] Invalid checkout form fields with when trying to purchase with PayPal (missing required field)
  - [x]   working with legacy mode enabled
- [x] Invalid checkout form fields when trying to purchase with the Debit Card option
  - [x] working with legacy mode enabled
- [x] Invalid checkout fields when trying to purchase with PayPal Credit option
  - [x] working with legacy mode enabled
- [x] Cancel a payment at PayPal and make sure we're redirected back to the cart page
- [x] Test checkout purchase with IE 11 - I only tested the basic checkout functionality on IE 11

#### Cart Page
- [x] SPB is not displayed when "Enable PayPal Checkout on the cart page" is disabled
- [x] Purchase with PayPal
- [x] Purchase with Debit Card
- [x] Purchase with PayPal Credit
- [x] Customer is redirected to "Confirm your PayPal Order" page
- [x] All the above are working with legacy mode enabled

#### Product Page
- [x] SPB not displayed when "Checkout on Single Product" is disabled
- [x] Product page-specific settings used only on the product page and not in mini-cart
- [x] Purchase with PayPal
- [x] Purchase with Debit Card
- [x] Purchase with PayPal Credit
- [x] Customer is redirected to "Confirm your PayPal Order" page
- [x] All the above are working with legacy mode enabled

#### Mini-Cart
- [x] SPB is not displayed when "Enable PayPal Checkout on the cart page" is disabled
- [x] Mini-cart specific settings used on mini-cart
- [x] Purchase with PayPal
- [x] Purchase with Debit Card
- [x] Purchase with PayPal Credit
- [x] Purchase through mini-cart while on a different product page
- [x] Customer is redirected to "Confirm your PayPal Order" page

**Note:** Something I noticed while testing is the debit card fields don't load in the mini-cart when the legacy mode is enabled and clicking the VISA button just opens the PayPal modal. With the new SDK however, the card fields are loaded: https://d.pr/i/xJmL01

I checked out `master` and to test what it used to do and it looks like the card fields never loaded in the mini-cart.

#### Merchant
- [x] Check orders in the merchant dashboard
